### ISSUE TITLE
Incredibly slow single-file db-update

### ISSUE THREAD

**Issue by aragaer** [org-roam/org-roam#1752]
### Saving a single large file takes seconds to update the database

Some of my files take literally seconds (14-15 seconds) to run org-roam-db-update file. At the same time if I disable update-on-save and just run org-roam-db-sync it takes about 0.7 seconds to process the same file.

I have tried to debug the issue myself but failed to get any significant results. Unfortunately I am not able to share a file that causes issue (but I have more than one of those) and I failed to reproduce the issue artificially. It does seem that the main factor is the number of org-headings coupled with org-links.

```
- command-execute                                                 605  70%
 - call-interactively                                             605  70%
  - funcall-interactively                                         518  59%
   - save-buffer                                                  468  54%
    - basic-save-buffer                                           468  54%
     - run-hooks                                                  464  53%
      - org-roam-db--update-on-save                               417  48%
       - org-roam-db-update-file                                  417  48%
        - org-roam-db-map-links                                   287  33%
         - org-element-map                                        172  19%
          - #<compiled 0x15679a850b9d>                            172  19%
           - mapc                                                 172  19%
            - #<compiled 0x15679a850b9d>                          172  19%
             - mapc                                               172  19%
              - #<compiled 0x15679a850b9d>                        172  19%
               - mapc                                             172  19%
                - #<compiled 0x15679a850b9d>                      172  19%
                 - mapc                                            97  11%
                  - #<compiled 0x15679a850b9d>                     97  11%
                   - mapc                                          97  11%
                    - #<compiled 0x15679a850b9d>                   97  11%
                     - mapc                                        87  10%
                      + #<compiled 0x15679a850b9d>                 87  10%
                     - #<compiled 0x15679a8e11d5>                  10   1%
                      - org-roam-db-insert-link                    10   1%
                       + org-get-outline-path                       9   1%
                       + org-roam-id-at-point                       1   0%
                 - #<compiled 0x15679a850b9d>                      73   8%
                  - mapc                                           73   8%
                   - #<compiled 0x15679a850b9d>                    73   8%
                    - #<compiled 0x15679a8e11d5>                   73   8%
                     - org-roam-db-insert-link                     73   8%
                      - org-roam-db-query                          66   7%
                       - apply                                     66   7%
                        - emacsql                                  66   7%
                         + apply                                   66   7%
                      + org-get-outline-path                        4   0%
                      + org-roam-id-at-point                        3   0%
                 + #<compiled 0x15679a8e11d5>                       2   0%
         - org-element-parse-buffer                               115  13%
          - org-element--parse-elements                           115  13%
           - org-element--parse-elements                          111  12%
            - org-element--parse-elements                         101  11%
             - org-element--parse-elements                         94  10%
              - org-element--parse-elements                        76   8%
               - org-element--parse-elements                       49   5%
                - org-element--parse-elements                      34   3%
                 - org-element--parse-elements                     18   2%
                  + org-element--parse-elements                     6   0%
                  + org-element--current-element                    4   0%
                  + org-element--parse-objects                      3   0%
                 + org-element--current-element                     9   1%
                 - org-element--parse-objects                       5   0%
                  + org-element--object-lex                         5   0%
                + org-element--current-element                      8   0%
                + org-element--parse-objects                        1   0%
               - org-element--current-element                      19   2%
                + org-element--list-struct                          9   1%
                  org-get-limited-outline-regexp                    3   0%
                  org-element-item-parser                           1   0%
                + org-element-headline-parser                       1   0%
               + org-element--parse-objects                         4   0%
              + org-element--current-element                       14   1%
              + org-element--parse-objects                          2   0%
             + org-element--current-element                         7   0%
            + org-element--current-element                          8   0%
           + org-element--current-element                           3   0%
        - org-roam-db-map-nodes                                   122  14%
         - org-map-entries                                        122  14%
          - org-scan-tags                                         112  12%
           - #<compiled 0x15679a8b65c9>                           107  12%
            + org-roam-db-insert-node-data                         64   7%
            + org-roam-db-node-p                                   26   3%
            + org-roam-db-insert-tags                               9   1%
            + org-roam-db-insert-aliases                            7   0%
            + org-roam-db-insert-refs                               1   0%
          + org-agenda-prepare-buffers                              9   1%
        + org-roam-db-insert-file                                   3   0%
        + org-roam-db-insert-file-node                              2   0%
          org-roam-db--file-hash                                    1   0%
        + org-set-regexps-and-options                               1   0%
        + org-roam-db-clear-file                                    1   0%
      - org-roam-link-replace-all                                  47   5%
       - org-roam-link-replace-at-point                            43   4%
        - org-element-context                                      38   4%
         - org-element-at-point                                    29   3%
          - org-element--parse-to                                  26   3%
           + org-element--current-element                          23   2%
             org-get-limited-outline-regexp                         1   0%
             outline-next-heading                                   1   0%
            org-get-limited-outline-regexp                          1   0%
          + org-at-heading-p                                        1   0%
         + org-element--object-lex                                  7   0%
          org-in-regexp                                             4   0%
     + basic-save-buffer-1                                          4   0%
   + file-notify-handle-event                                      23   2%
   + evil-normal-state                                             16   1%
   + execute-extended-command                                      10   1%
  + byte-code                                                      87  10%
+ ...                                                             245  28%
+ timer-event-handler                                              11   1%
+ redisplay_internal (C function)                                   2   0%
  internal-echo-keystrokes-prefix                                   1   0%

```

- Emacs: GNU Emacs 27.2 (build 1, x86_64-pc-linux-gnu, X toolkit)
 of 2021-07-26
- Framework: N/A
- Org: Org mode version 9.4.6 (9.4.6-12-gdcc3a8-elpaplus @ /home/aragaer/.config/emacs/elpa/org-plus-contrib-20210802/)
- Org-roam: 2.0.0

**Comment by naistran** [comment]
Could you try this advice and see if it's faster?

```elisp
(advice-add 'org-roam-db-update-file :around
              (defun +org-roam-db-update-file (fn &rest args)
                  (emacsql-with-transaction (org-roam-db)
                    (apply fn args))))
```

This should batch all SQLite operations resulting from `org-roam-db-update-file` in one transaction.

**Comment by aragaer** [comment]
This way it's blazing fast.

**Comment by PRESFIL** [comment]
It seems like `emacsql-with-transaction` is already in use - and it didn't make
any difference to me either, or I couldn't properly use it with `use-package`.

I also found an interesting thing: for me, a table with **80 inactive
timestamps** contributed more to the performance drop. As soon as I removed it
from the file, it worked instantly. The original file contains a large table
with timestamps and 120 headers. With this table, emacs would hang for 15
seconds. Only with 120 headers - fast.

**Comment by xbc5** [comment]
> Could you try this advice and see if it's faster?
> 
> ```emacs-lisp
> (advice-add 'org-roam-db-update-file :around
>               (defun +org-roam-db-update-file (fn &rest args)
>                   (emacsql-with-transaction (org-roam-db)
>                     (apply fn args))))
> ```
> 
> This should batch all SQLite operations resulting from `org-roam-db-update-file` in one transaction.

It works for me. Thanks!

**Comment by zecevicp** [comment]
Does the file have a lot of http and file links? In that case I think you can speed it up with the fix described in #1824.

**Comment by zecevicp** [comment]
@Wetlize  Can you please elaborate why is it necessary for roam to store http, https, file and other links to its database? You said in #1824 that it's a feature and I wouldn't want to miss on anything... Thanks

**Comment by Wetlize** [comment]
@zecevicp Sure, we expose this information a separate `links` table from our DB schemata:
https://github.com/org-roam/org-roam/blob/1795039ab93ef19611dbb3fca21c4211c4e655a9/org-roam-db.el#L189-L195
for extensions, ourselves and the end users, with which you can query to the database all the nodes with the related links (and vice versa) for various purposes. In Org-roam itself we use this table for:
1. The `org-roam-buffer` to get all the implicit links between nodes that link by referencing assigned `ROAM_REFS` and then extracting the preview content for them in `org-roam-reflinks-section`. Obviously, the same table is used for `org-roam-backlinks-section` to do the same thing for explicit `id` links between nodes.
2. The `org-roam-graph` to display forward links from each node as unique vertices.
    - I believe `org-roam-ui` can also implement this functionality too, but I'm not sure whether they do so.

All of which isn't limited in terms of purpose and can be used for other too, that just not implemented ATM or wasn't thought about.

**Comment by zecevicp** [comment]
OK, thanks. I think I can stick with only storing id links for now.

**Comment by rasendubi** [comment]
I also experience extremely slow saving with a big file. It has some links and timestamps, no headings, no external links. [The advice above](https://github.com/org-roam/org-roam/issues/1752#issuecomment-896074821) doesn’t help. org-roam-20210901.1143

I have debugged it to a minimal reproducible example which is just a ton of links and a ton of timestamps. If I delete either links or timestamps, it’s fast again.

<details>
<summary><b>Reproducible example</b></summary>

```org
:PROPERTIES:
:ID:       20210924071313
:END:
#+TITLE: title

[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]

[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]

[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]

[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]

[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]
[[id:xxx][link]]

[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]

[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]

[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]

[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]

[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
[2021-09-24 Fri]
```

</details>

**Comment by natejlong** [comment]
I'm also affected by this. I've just updated to the latest org-roam after not taking any update for the last month or two, and now I'm seeing 4-5s save times on my main todo files (lots of headings and timestamps, <30 links)

From https://github.com/org-roam/org-roam/issues/1824 it sounds like a solution is planned? 
> There's solution that we have for #1752, but it requires from us to get some other things done before we can get to it. In any case, I'm closing this, since the issue is clearly a duplicate of #1752, so please, keep all the related things there.

Do you all want any more debug info or is it just a matter of being patient until that solution gets rolled out?

**Comment by rasendubi** [comment]
From the discussion above, it doesn’t seem like there is a solution planned. The only solution mentioned is wrapping org-roam-db-update-file into a transaction and it doesn’t help with the multiple timestamps multiple links case.

There is a repro case now, so I don’t think more input from the users will help (unless that input is a new repro case with a different pattern).

The next step should be for someone to try and find the root cause of the issue. I briefly tried to debug it but didn’t succeed, so volunteers are welcome.

**Comment by naistran** [comment]
From my profiling, when there are a lot of timestamps then `org-entry-properties` (used by `org-roam-db-node-p`) is slow. I tried `org-entry-get` which should do what we need to check for "ROAM_EXCLUDE" but is a lot faster:

```lisp
(advice-add 'org-roam-db-node-p :override
            (defun +org-roam-db-node-p ()
              (and (org-id-get)
                   (not (org-entry-get (point) "ROAM_EXCLUDE"))
                   (funcall org-roam-db-node-include-function))))
```

**Comment by PRESFIL** [comment]
@naistran , I totally agree about timestamps. Also, parsing of timestamps inside tables is twice as slow :smiley: .

Also, the recent PR #1868 for processing links inside PROPERTIES brings its 10 cents into the slowdown.

**Comment by naistran** [comment]
I haven't tested that commit yet. To clarify, you're saying it makes saving even slower?

**Comment by PRESFIL** [comment]
Yes, [this](https://github.com/org-roam/org-roam/issues/1752#issuecomment-897092800) file takes ~26 seconds to save now.
So there are file that saved almost instantly. Now it is saved for about 2 seconds.

**Comment by bhipple** [comment]
I have a large org-mode table with inactive timestamps inside logging gym workouts and what I did going back many years.  I recently moved it into my org-roam folder and db updates became cripplingly slow (60+ **minutes** on Linux).  

Today I updated to the latest commit from yesterday and I'm back down to doing a db sync in ~2 seconds!  Thank you for taking performance seriously; this is great!

**Comment by arbox** [comment]
I'm experiencing a very long DB update on files with longer logging history (like one entry per day in the `:LOGBOOK:` drawer for the last 7 years):
```
:LOGBOOK:
CLOCK: [2021-11-22 Mon 14:15]
CLOCK: [2021-11-19 Fri 12:59]--[2021-11-19 Fri 13:14] =>  0:15
CLOCK: [2021-11-15 Mon 16:10]--[2021-11-15 Mon 16:27] =>  0:17
CLOCK: [2021-11-15 Mon 15:22]--[2021-11-15 Mon 15:46] =>  0:24
CLOCK: [2021-11-15 Mon 12:52]--[2021-11-15 Mon 12:53] =>  0:01
CLOCK: [2021-11-15 Mon 11:28]--[2021-11-15 Mon 12:28] =>  1:00
CLOCK: [2021-11-12 Fri 14:21]--[2021-11-12 Fri 15:03] =>  0:42
CLOCK: [2021-11-12 Fri 14:03]--[2021-11-12 Fri 14:04] =>  0:01
CLOCK: [2021-11-12 Fri 12:14]--[2021-11-12 Fri 13:14] =>  1:00
CLOCK: [2021-11-12 Fri 11:01]--[2021-11-12 Fri 11:21] =>  0:20
CLOCK: [2021-11-11 Thu 15:59]--[2021-11-11 Thu 16:46] =>  0:47
CLOCK: [2021-11-11 Thu 09:27]--[2021-11-11 Thu 10:02] =>  0:35
- State "DOING"      from "TODO"       [2021-11-10 Wed 17:36]
CLOCK: [2021-11-10 Wed 17:36]--[2021-11-10 Wed 17:51] =>  0:15
CLOCK: [2021-11-10 Wed 17:03]--[2021-11-10 Wed 17:21] =>  0:18
CLOCK: [2021-11-10 Wed 16:25]--[2021-11-10 Wed 16:26] =>  0:01
CLOCK: [2021-11-10 Wed 14:54]--[2021-11-10 Wed 15:44] =>  0:50
CLOCK: [2021-11-10 Wed 13:53]--[2021-11-10 Wed 13:59] =>  0:06
CLOCK: [2021-11-10 Wed 12:14]--[2021-11-10 Wed 12:52] =>  0:38
CLOCK: [2021-11-09 Tue 17:11]--[2021-11-09 Tue 17:24] =>  0:13
CLOCK: [2021-11-09 Tue 10:59]--[2021-11-09 Tue 11:02] =>  0:03
CLOCK: [2021-11-09 Tue 10:47]--[2021-11-09 Tue 10:56] =>  0:09
CLOCK: [2021-11-09 Tue 10:46]--[2021-11-09 Tue 10:47] =>  0:01
# Removed other ~ 2000 entries.
:END:
```

**Comment by salutis** [comment]
I have two large Org files, one with 40,868 lines and the other with 4,842 lines.  A couple of days ago, I migrated the _smaller_ file (with just 687 headings) from plain Org to Org Roam, and it made Emacs incredibly slow, resulting in e.g. 10-second lockups on every save.  I guess Org Roam has zero chances of handling the larger file?

**Comment by salutis** [comment]
**Note:**  Yesterday, I split one of my big Org files into about 450 small files and Org Roam works _super-great_ now.  Fast!
